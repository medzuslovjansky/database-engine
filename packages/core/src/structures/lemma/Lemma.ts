import type { SteenbergenLemmaMetadata } from './SteenbergenLemmaMetadata';
import { parseLemma } from './parseLemma';

export type LemmaOptions = {
  value: string;
  annotations: string[];
};

export class Lemma {
  constructor(options: Partial<LemmaOptions> = {}) {
    this.value = options.value ?? '';
    this.annotations = options.annotations ?? [];
  }

  public value: string;

  public annotations: string[];

  public steen?: SteenbergenLemmaMetadata;

  public clone(): Lemma {
    return new Lemma({
      value: this.value,
      annotations: [...this.annotations],
    });
  }

  public toString(): string {
    if (this.annotations.length === 0) {
      return this.value;
    }

    return `${this.value} (${this.annotations.map(String).join('; ')})`;
  }

  public static parse(str: string): Lemma {
    return new Lemma(parseLemma(str));
  }
}
