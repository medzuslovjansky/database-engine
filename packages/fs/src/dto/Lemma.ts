import type { Language } from './misc';

export type Lemma = {
  id: number;
  isv: string[];
  partOfSpeech: string;
  translations: Partial<Record<Language, string[]>>;
  examples?: string[];
  type?: number;
  sameInLanguages?: string;
  genesis?: string;
  frequency?: number;
  intelligibility?: string;
  // metadata: {
  //   translations: TranslationMetadata;
  // };
};

// export type TranslationMetadata = {
//   autotranslated: Set<Language>;
//   debatable: Set<Language>;
// };
//
// export type Intelligibility = {
//   [key in SlavicLanguage]?: IntelligibilityStatus;
// };
//
// export type IntelligibilityStatus = '+' | '~' | '-';
