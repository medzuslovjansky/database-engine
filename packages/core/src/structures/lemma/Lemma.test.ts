import { Lemma } from './Lemma';

describe('Lemma', () => {
  let lemma: Lemma;

  describe('when created', () => {
    describe('with no args', () => {
      beforeEach(() => {
        lemma = new Lemma();
      });

      it('should have an empty value', () => {
        expect(lemma.value).toBe('');
      });

      it('should have no annotations', () => {
        expect(lemma.annotations).toEqual([]);
      });
    });

    describe('parsed from a string', () => {
      beforeEach(() => {
        lemma = Lemma.parse('');
      });

      test.each([
        'běguči',
        'око (устар.)',
        'той, що біжить',
        'bring (up) on (colloq.; slang)',
      ])('%j should be stringified to the same form', (testString) => {
        const lemma = Lemma.parse(testString);
        expect(lemma).toMatchSnapshot();
        expect(`${lemma}`).toBe(testString);
      });

      it('should ignore empty annotations', () => {
        const lemma = Lemma.parse('pråzdne anotacije ()');
        expect(lemma.value).toBe('pråzdne anotacije');
        expect(lemma.annotations).toEqual([]);
        expect(`${lemma}`).toBe('pråzdne anotacije');
      });

      test.each([
        'zabezpamečena lěva zatvorka)',
        'zabezpamečena prava zatvorka (',
        'měšane ) zatvorky (',
      ])(
        'should throw an error if there are unbalanced brackets',
        (testString) => {
          expect(() => Lemma.parse(testString)).toThrow(
            /incorrect parentheses/,
          );
        },
      );
    });

    describe('with value and annotations', () => {
      beforeEach(() => {
        lemma = new Lemma({
          value: 'this',
          annotations: ['demonstrative'],
        });
      });

      it('should have that value', () => {
        expect(lemma.value).toBe('this');
      });

      it('should have those annotations', () => {
        expect(lemma.annotations).toEqual(['demonstrative']);
      });
    });
  });

  describe('when stringified', () => {
    it('should convert into its value if there are no annotations', () => {
      lemma = new Lemma({ value: 'and' });
      expect(`${lemma}`).toBe('and');
    });

    it('should convert into its value with annotations separated by ";" inside round brackets', () => {
      lemma = new Lemma({
        value: 'and',
        annotations: ['conj.', 'common'],
      });

      expect(`${lemma}`).toBe('and (conj.; common)');
    });
  });
});
