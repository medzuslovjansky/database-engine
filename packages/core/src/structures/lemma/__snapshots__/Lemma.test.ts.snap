// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lemma when created parsed from a string "bring (up) on (colloq.; slang)" should be stringified to the same form 1`] = `
Lemma {
  "annotations": [
    "colloq.",
    "slang",
  ],
  "value": "bring (up) on",
}
`;

exports[`Lemma when created parsed from a string "běguči" should be stringified to the same form 1`] = `
Lemma {
  "annotations": [],
  "value": "běguči",
}
`;

exports[`Lemma when created parsed from a string "око (устар.)" should be stringified to the same form 1`] = `
Lemma {
  "annotations": [
    "устар.",
  ],
  "value": "око",
}
`;

exports[`Lemma when created parsed from a string "той, що біжить" should be stringified to the same form 1`] = `
Lemma {
  "annotations": [],
  "value": "той, що біжить",
}
`;
